### ISSUE TITLE
Errors opening unit ontology with Protégé

### ISSUE THREAD

**Issue by yoavnash** [qudt/qudt-public-repo#62]
When I try to open the [unit ontology](http://www.qudt.org/2.1/vocab/unit) with Protégé 5.5, I get the following errors:
```
...
   INFO  16:15:32  Usage of rdf vocabulary: http://www.linkedmodel.org/schema/dtype#value -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Property
   INFO  16:15:32  Usage of rdf vocabulary: http://www.linkedmodel.org/schema/dtype#value -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Property
   INFO  16:15:32  Usage of rdf vocabulary: http://www.linkedmodel.org/schema/dtype#defaultValue -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Property
  ERROR  16:15:32  Entity not properly recognized, missing triples in input? http://org.semanticweb.owlapi/error#Error1 for type Class
  ERROR  16:15:32  Entity not properly recognized, missing triples in input? http://org.semanticweb.owlapi/error#Error2 for type Class
   INFO  16:15:32  Unparsed triple: _:genid2147483682 -> http://www.w3.org/2002/07/owl#onProperty -> http://www.linkedmodel.org/schema/dtype#defaultValue
   INFO  16:15:32  Unparsed triple: _:genid2147483685 -> http://www.w3.org/2002/07/owl#onProperty -> http://www.linkedmodel.org/schema/dtype#value
   INFO  16:15:32  Finished loading imported ontology at http://www.linkedmodel.org/schema/dtype
...
   INFO  16:11:07  Usage of rdf vocabulary: http://www.w3.org/2004/02/skos/core#exactMatch -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Property
   INFO  16:11:07  Usage of rdf vocabulary: http://www.w3.org/2004/02/skos/core#closeMatch -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Property
   INFO  16:11:07  Finished loading imported ontology at http://www.w3.org/2004/02/skos/core
  ERROR  16:11:07  Entity not properly recognized, missing triples in input? http://org.semanticweb.owlapi/error#Error3 for type Class
  ERROR  16:11:07  Entity not properly recognized, missing triples in input? http://org.semanticweb.owlapi/error#Error4 for type Class
  ERROR  16:11:07  Entity not properly recognized, missing triples in input? http://org.semanticweb.owlapi/error#Error5 for type Class
  ERROR  16:11:07  Entity not properly recognized, missing triples in input? http://org.semanticweb.owlapi/error#Error6 for type Class
...
   INFO  16:15:34  Imported ontology document http://qudt.org/2.1/vocab/quantitykind was not resolved to any documents defined in the ontology catalog.
   INFO  16:15:35  Imported ontology document http://qudt.org/2.1/vocab/dimensionvector was not resolved to any documents defined in the ontology catalog.
  ERROR  16:15:37  Attempt to transform an axiom to correct misuse of properties failed. Property replacement: {<http://www.linkedmodel.org/schema/vaem#hasDomain>=<http://www.linkedmodel.org/schema/vaem#hasDomain>, <http://qudt.org/schema/qudt/ucumCaseSensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseSensitiveCode>, <http://qudt.org/schema/qudt/rationale>=<http://qudt.org/schema/qudt/rationale>, <http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>, <http://qudt.org/schema/qudt/vectorMagnitude>=<http://qudt.org/schema/qudt/vectorMagnitude>}, axiom: AnnotationAssertion(<http://www.linkedmodel.org/schema/vaem#hasDomain> <http://www.linkedmodel.org/schema/vaem#GMD_QUDT-DIMENSION-VECTORS> "dimensionVector"), error: uk.ac.manchester.cs.owl.owlapi.OWLLiteralImplNoCompression cannot be cast to org.semanticweb.owlapi.model.IRI
   INFO  16:15:37  Finished loading imported ontology at http://qudt.org/2.1/vocab/dimensionvector
   INFO  16:15:37  Imported ontology document http://qudt.org/2.1/vocab/unit was not resolved to any documents defined in the ontology catalog.
  ERROR  16:15:39  Attempt to transform an axiom to correct misuse of properties failed. Property replacement: {<http://www.linkedmodel.org/schema/vaem#hasDomain>=<http://www.linkedmodel.org/schema/vaem#hasDomain>, <http://qudt.org/schema/qudt/ucumCaseSensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseSensitiveCode>, <http://qudt.org/schema/qudt/rationale>=<http://qudt.org/schema/qudt/rationale>, <http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>, <http://qudt.org/schema/qudt/vectorMagnitude>=<http://qudt.org/schema/qudt/vectorMagnitude>}, axiom: AnnotationAssertion(<http://www.linkedmodel.org/schema/vaem#hasDomain> <http://www.linkedmodel.org/schema/vaem#GMD_QUDT-DIMENSION-VECTORS> "dimensionVector"), error: uk.ac.manchester.cs.owl.owlapi.OWLLiteralImplNoCompression cannot be cast to org.semanticweb.owlapi.model.IRI
   INFO  16:15:39  Finished loading imported ontology at http://qudt.org/2.1/vocab/unit
  ERROR  16:15:39  Attempt to transform an axiom to correct misuse of properties failed. Property replacement: {<http://www.linkedmodel.org/schema/vaem#hasDomain>=<http://www.linkedmodel.org/schema/vaem#hasDomain>, <http://qudt.org/schema/qudt/ucumCaseSensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseSensitiveCode>, <http://qudt.org/schema/qudt/rationale>=<http://qudt.org/schema/qudt/rationale>, <http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>, <http://qudt.org/schema/qudt/vectorMagnitude>=<http://qudt.org/schema/qudt/vectorMagnitude>}, axiom: AnnotationAssertion(<http://www.linkedmodel.org/schema/vaem#hasDomain> <http://www.linkedmodel.org/schema/vaem#GMD_QUDT-QUANTITY-KINDS-ALL> "quantitykind"), error: uk.ac.manchester.cs.owl.owlapi.OWLLiteralImplNoCompression cannot be cast to org.semanticweb.owlapi.model.IRI
  ERROR  16:15:39  Attempt to transform an axiom to correct misuse of properties failed. Property replacement: {<http://www.linkedmodel.org/schema/vaem#hasDomain>=<http://www.linkedmodel.org/schema/vaem#hasDomain>, <http://qudt.org/schema/qudt/ucumCaseSensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseSensitiveCode>, <http://qudt.org/schema/qudt/rationale>=<http://qudt.org/schema/qudt/rationale>, <http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>, <http://qudt.org/schema/qudt/vectorMagnitude>=<http://qudt.org/schema/qudt/vectorMagnitude>}, axiom: AnnotationAssertion(<http://www.linkedmodel.org/schema/vaem#hasDomain> <http://www.linkedmodel.org/schema/vaem#GMD_QUDT-DIMENSION-VECTORS> "dimensionVector"), error: uk.ac.manchester.cs.owl.owlapi.OWLLiteralImplNoCompression cannot be cast to org.semanticweb.owlapi.model.IRI
   INFO  16:15:39  Finished loading imported ontology at http://qudt.org/2.1/vocab/quantitykind
  ERROR  16:15:39  Attempt to transform an axiom to correct misuse of properties failed. Property replacement: {<http://www.linkedmodel.org/schema/vaem#hasDomain>=<http://www.linkedmodel.org/schema/vaem#hasDomain>, <http://qudt.org/schema/qudt/ucumCaseSensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseSensitiveCode>, <http://qudt.org/schema/qudt/rationale>=<http://qudt.org/schema/qudt/rationale>, <http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>, <http://qudt.org/schema/qudt/vectorMagnitude>=<http://qudt.org/schema/qudt/vectorMagnitude>}, axiom: AnnotationAssertion(<http://www.linkedmodel.org/schema/vaem#hasDomain> <http://www.linkedmodel.org/schema/vaem#GMD_QUDT-QUANTITY-KINDS-ALL> "quantitykind"), error: uk.ac.manchester.cs.owl.owlapi.OWLLiteralImplNoCompression cannot be cast to org.semanticweb.owlapi.model.IRI
  ERROR  16:15:39  Attempt to transform an axiom to correct misuse of properties failed. Property replacement: {<http://www.linkedmodel.org/schema/vaem#hasDomain>=<http://www.linkedmodel.org/schema/vaem#hasDomain>, <http://qudt.org/schema/qudt/ucumCaseSensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseSensitiveCode>, <http://qudt.org/schema/qudt/rationale>=<http://qudt.org/schema/qudt/rationale>, <http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>=<http://qudt.org/schema/qudt/ucumCaseInsensitiveCode>, <http://qudt.org/schema/qudt/vectorMagnitude>=<http://qudt.org/schema/qudt/vectorMagnitude>}, axiom: AnnotationAssertion(<http://www.linkedmodel.org/schema/vaem#hasDomain> <http://www.linkedmodel.org/schema/vaem#GMD_QUDT-DIMENSION-VECTORS> "dimensionVector"), error: uk.ac.manchester.cs.owl.owlapi.OWLLiteralImplNoCompression cannot be cast to org.semanticweb.owlapi.model.IRI
   INFO  16:15:39  Finished loading file:/C:/Users/Yoav/Downloads/qudt/unit
   INFO  16:15:39  Loading for ontology and imports closure successfully completed in 10067 ms
   INFO  16:15:39  Updated document format class from: org.semanticweb.owlapi.formats.RioTurtleDocumentFormat to: org.semanticweb.owlapi.formats.TurtleDocumentFormat
   INFO  16:15:39  
...
```
Am I doing something wrong?

**Comment by steveraysteveray** [comment]
Thanks for the error messages. 
QUDT has been developed using the TopQuadrant suite of ontology editing tools. However, you aren't the first person to note that Protege gives errors when loading QUDT.
I will try to spend a little time to better understand what it is that Protege is complaining about.

**Comment by smrgeoinfo** [comment]
related: https://github.com/ESIPFed/sweet/issues/84, https://github.com/owlcs/owlapi/issues/571  

I got the QUDT ttl files to load in Protege without all the bogus error classes by :

1. download local copies of http://www.linkedmodel.org/schema/dtype and http://www.linkedmodel.org/schema/vaem
2. checking out the qudt/qudt-public-repo repository from github
3. creating oasis catalog files in the schema and vocab directories and subdirectories to use local copies of all the ontology imports; have to generate catalog files in directories and subdirectories until all the imports except shex, shacl, spin are local.
4. in protege manually delete the Error classes when I load one of the files, then save.  I did this for vaem, imports, dtype, qudt, data, datatype, units, quantity-kind. After saving, the ontology will reopen without the Error classes. 

Since protege rewrites all the triples in different order, I can't tell what if any changes were actually made in the .ttl files when saved... - git diff shows the whole file being different.

**Comment by steveraysteveray** [comment]
Thanks so much for diagnosing the problem here! It's good to know the root cause. Interesting that the imports must be local...
We will discuss.

**Comment by jhodgesatmb** [comment]
When I load the unit ontology into Protege 5.5 I get many errors but I don't know how to interpret them in terms of cause. I see that the offending QUDT class, property, or instance is subclassed off a Protege Error Class. The scant Protege documentation says to look at the usage of the Error Class to find out what the problem is but I am not finding any help there.

In order to fix these errors I need to be able to understand what these errors really mean. When @smrgeoinfo says he downlopaded the dtype and vaem graphs, how was it that he decided this was necessary when the downloaded version would have the same content as the networked version? When he says he created catalog files in the schema and vocab directories, what was the reasoning behind that and what errors did this action resolve? It seems that he had issues with non-local imports, and I have run into this kind of problem in the past but I would like to hear what the problem was perhaps more than the solution.

By the way, how long did it take to perform step #3? I could see this taking quite some time. Once I find out what the problems are we can discuss whether and how to resolve them.

**Comment by dr-shorthair** [comment]
My experience is that Protégé throws these a lot for 
(a) owl:Restriction classes, where 
(b) Dublin Core is involved, which is 'RDF' and not very good OWL. 
Protégé is very much an OWL tool, and requires OWL consistency, while TopBraid is more RDF/RDFS oriented.

**Comment by dr-shorthair** [comment]
And as @smrgeoinfo points out, the RDF serialiser in TopBraid writes the file in a different order to OWL-API which is used by Protege so it is difficult to run diffs. If you want to use GitHub diffs, then you need to choose one of these as the local standard, and rewrite the file using the locally preferred tool first. It could probably be automated with a post-commit hook, but I don't have the skill to do that.

**Comment by GeorgFerdinandSchneider** [comment]
Hi I have a potentially related or similar error from opening qudt-s: <http://qudt.org/schema/qudt/> in Protege 5.5 from the url and then trying to execute HermiT reasoner on it:

´´´
 ERROR  17:24:34  An error occurred during reasoning: Non-simple property '<http://qudt.org/schema/qudt/isScalingOf>' or its inverse appears in the cardinality restriction 'ObjectMaxCardinality(1 <http://qudt.org/schema/qudt/isScalingOf> owl:Thing)'..
java.lang.IllegalArgumentException: Non-simple property '<http://qudt.org/schema/qudt/isScalingOf>' or its inverse appears in the cardinality restriction 'ObjectMaxCardinality(1 <http://qudt.org/schema/qudt/isScalingOf> owl:Thing)'.
	at org.semanticweb.HermiT.structural.ObjectPropertyInclusionManager.rewriteAxioms(ObjectPropertyInclusionManager.java:125) ~[na:na]
	at org.semanticweb.HermiT.structural.OWLClausification.preprocessAndClausify(OWLClausification.java:161) ~[na:na]
	at org.semanticweb.HermiT.Reasoner.loadOntology(Reasoner.java:211) ~[na:na]
	at org.semanticweb.HermiT.Reasoner.<init>(Reasoner.java:202) ~[na:na]
	at org.semanticweb.HermiT.Reasoner.<init>(Reasoner.java:176) ~[na:na]
	at org.semanticweb.HermiT.ReasonerFactory.createHermiTOWLReasoner(ReasonerFactory.java:51) ~[na:na]
	at org.semanticweb.HermiT.ReasonerFactory.createReasoner(ReasonerFactory.java:19) ~[na:na]
	at org.protege.editor.owl.model.inference.ReasonerUtilities.createReasoner(ReasonerUtilities.java:20) ~[na:na]
	at org.protege.editor.owl.model.inference.OWLReasonerManagerImpl$ClassificationRunner.ensureRunningReasonerInitialized(OWLReasonerManagerImpl.java:428) ~[na:na]
	at org.protege.editor.owl.model.inference.OWLReasonerManagerImpl$ClassificationRunner.run(OWLReasonerManagerImpl.java:386) ~[na:na]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_121]
´´´
Basically I like to evaluate which OWL profile QUDT is belonging to. @steveraysteveray  did you play around with different OWL-DL reasoners sofar?

**Comment by steveraysteveray** [comment]
As it happens, I'm using the SHACL version of QUDT in another project where we use the OWL2 RL profile. I'm currently working through some issues that came up regarding the implications of owl:sameAs that seems to be replicating what I thought were annotation properties, and thus violating some cardinality constraints of 1. 
Having said that, QUDT doesn't "belong" to any OWL profile, at least not intentionally!

**Comment by vChavezB** [comment]
Another protege user here. I tried opening the QuantityKind vocabulary but got these errors when reasoning with Pellet:


## rdf:html data type parsing

Reason for inconsistency: Literal value "`QUDT Dimension Vectors is a vocabulary that extends QUDT Quantity Kinds with properties that support dimensional analysis. There is one dimension vector for each of the system\'s base quantity kinds. The vector\'s magnitude determines the exponent of the base dimension for the referenced quantity kind`"^^HTML does not belong to datatype string

What I did is deleted the rdf:html data type and this error was removed.

## quantityKind:SoundExposureLevel restrictions

Reason for inconsistency: Individual http://qudt.org/vocab/quantitykind/SoundExposureLevel has more than 1 values for property http://qudt.org/schema/qudt/latexDefinition violating the cardinality restriction

I deleted from the the `SoundExposureLevel` individual the data property `latexDefinition` but I still got the error. So I also deleted its other data property `symbol` and now it works.


Do you have any other suggestion on how the vocabulary could be loaded correctly for reasoning?

I tried only loading the qudt schema directly over the URL `http://qudt.org/2.1/schema/qudt` with protege and reasoning works but with the vocabulary QuantityKinds Im having these problems.

**Comment by steveraysteveray** [comment]
We are looking into this. Thanks for pointing this out.

**Comment by vChavezB** [comment]
So far the problem I mentioned in the previous https://github.com/qudt/qudt-public-repo/issues/62#issuecomment-1781813198 works fine with Pellet :)

**Comment by Toby-Broom** [comment]
I also don't have any issues with Protege

**Comment by fkleedorfer** [comment]
I think this is resolved @steveraysteveray, right?

**Comment by steveraysteveray** [comment]
Yes, resolved. Runs fine under Protege 5.6.4.

**Comment by steveraysteveray** [comment]
Sigh. I take your point on your first paragraph. In the meantime perhaps you could use the explicitly versioned 3.0.0, as in https://qudt.org/3.0.0/schema/qudt/, until the next Release (3.1.1) scheduled for April. That version does not have the qudt:hasFactorUnit relation.

Regarding the points raised in your second and third paragraphs, I know that Protege 5.6.4 loads without errors with https://qudt.org/2.1/schema/qudt, (or even https://qudt.org/2.1/vocab/unit) and the things you bring up were present then. Are you talking about errors here?

**Comment by steveraysteveray** [comment]
Let me amend my "meantime" suggestion, since 3.0.0 also seems to show errors in Protege. Need to use 2.1 for now.

**Comment by hentoilusose** [comment]
Sorry for my initial confusingly worded message - I did not mean the latter points were errors, they were merely pointed out as challenges to be overcome if QUDT is intended to be used with inferencing engines (preferably DL or narrower profile). The parsing issues stem from from OWL API that is widely shared, so e.g. validating with OBO ROBOT against OWL Full gives the same errors.

**Comment by fkleedorfer** [comment]
I've been working on the QUDT build process including automated checks, and it seems like these owlapi checks would be a welcome addition, given that you implicitly perform these checks when using QUDT. Not knowing owlapi so well, do you think the checks you are doing could be packed into a maven plugin?

**Comment by hentoilusose** [comment]
@fkleedorfer Yes! Even though i more often use it from the command line, I think ROBOT is by far the easiest to integrate as it builds on OWL API and is written in Java (as is the case also with the embedded inference engines): https://robot.obolibrary.org/

**Comment by fkleedorfer** [comment]
I'd like to include that in the build then. Could you provide a java class (maybe in a GitHub gist) that does the checks you want?

**Comment by hentoilusose** [comment]
Unfortunately I don't have time to give further assistance, but I recommend starting by checking this static method out: https://github.com/ontodev/robot/blob/2bc146f5a9a521c061c6f55cc282ed4b5e58a687/robot-core/src/main/java/org/obolibrary/robot/ReasonerHelper.java#L123

I hope this helps!

**Comment by steveraysteveray** [comment]
@hentoilusose, just to close the loop on this, I believe I have made changes to the OWL schema file so that it once more loads without error, at least with Protege 5.6.4.  These changes are reflected in the [latest snapshot release](https://github.com/qudt/qudt-public-repo/releases/tag/snapshot), but not yet deployed as resolvable URIs on the internet. That is scheduled for the last week of April, as part of Release 3.1.1.
